src/integrity-checker.js

```diff
@@ -33,30 +39,37 @@ export default class InstallationIntegrityChecker {
    * write a new one.
    */
 
-  async _getIntegrityHashLocation(): Promise<string> {
+  async _getIntegrityHashLocation(usedRegistries?: Set<RegistryNames>): Promise<IntegrityHashLocation> {
     // build up possible folders
+    let registries = registryNames;
+    if (usedRegistries && usedRegistries.size > 0) {
+      registries = usedRegistries;
+    }
     const possibleFolders = [];
     if (this.config.modulesFolder) {
       possibleFolders.push(this.config.modulesFolder);
     }
 
     // ensure we only write to a registry folder that was used
-    for (const name of registryNames) {
+    for (const name of registries) {
       const loc = path.join(this.config.cwd, this.config.registries[name].folder);
       possibleFolders.push(loc);
     }
 
     // if we already have an integrity hash in one of these folders then use it's location otherwise use the
     // first folder
     const possibles = possibleFolders.map((folder): string => path.join(folder, constants.INTEGRITY_FILENAME));
-    let loc = possibles[0];
+    let loc;
     for (const possibleLoc of possibles) {
       if (await fs.exists(possibleLoc)) {
         loc = possibleLoc;
         break;
       }
     }
```

Comment:
This script will return the first folder that has .yarn-integrity file.
If none found it will pick the first from the list of available.

Right now it is hard to decide whether existence of `bower_modules` is the right signal for Yarn to save .yarn-integrity there.
Let's get back to this subject when Yarn has more than one folder it can serve, currently it is node_modules only